### PR TITLE
Version pinning

### PR DESCRIPTION
**Pull request by ddunbar**

### CODE REVIEW

File: proposals/NNNN-swiftpm-version-pinning.md [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+	```
+	$ swift package update [--repin]
+	```
+
+	* Update command errors if there are no unpinned packages which can be updated.
+
+	* Otherwise, the behaviour is to update all unpinned packages to the latest possible versions which can be resolved while respecting the existing pins.
+
+	* The `[--repin]` argument can be used to lift the version pinning restrictions. In this case, the behaviour is that all packages are updated, and packages which were previously pinned are then repinned to the latest resolved versions.
```
Comment (BasThomas) [inline proposals/NNNN-swiftpm-version-pinning.md]: behaviour > behavior

File: proposals/NNNN-swiftpm-version-pinning.md [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+We have intentionally kept the pin file format an implementation detail in order
+to allow for future expansion. For example, we would like to consider embedding
+additional information on a known tag (like its SHA, when using Git) in the pins
+file as a security feature, to prevent man-in-the-middle attacks on parts of the
+package graph.
+
+## Impact on existing code
+
+There will be change in the behaviours of `swift build` and `swift package update` in presence of the pins file, as noted in the proposal however the existing package will continue to build without any modifications.
```
Comment (BasThomas) [inline proposals/NNNN-swiftpm-version-pinning.md]: There should be a comma between "proposal" and "however"

File: proposals/NNNN-swiftpm-version-pinning.md [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+Where other package managers will simultaneously link two versions
+of `LibX` -- and hope that the differing simultaneous uses of
+`LibX` do not cause other compile-time or run-time errors -- 
+SwiftPM will simply fail to resolve the dependencies.
+
+We therefore wish to encourage library authors to keep their
+packages building and testing with as recent and as wide a range
+of versions of their dependencies as possible, and guard more
+vigorously than other systems against accidental overconstraint.
+One way to encourage this behaviour is to avoid emitting pins files
```
Comment (BasThomas) [inline proposals/NNNN-swiftpm-version-pinning.md]: behavior

File: proposals/NNNN-swiftpm-version-pinning.md [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+by default.
+
+We also belief that if packages default to exposing their pin files as part of
```
Comment (BasThomas) [inline proposals/NNNN-swiftpm-version-pinning.md]: belief > believe

File: proposals/NNNN-swiftpm-version-pinning.md [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+1. We do not yet have a robust dependency resolution algorithm we can rely
+   on. The complexity of the algorithm is in some ways relative to the degree of
+   conflicts we expect to be present in the package graph (for example, this may
+   mean we need to investigate significantly more work in optimizing its
+   performance, or in managing its diagnostics).
+
+2. The Swift package manager and its ecosystem is evolving quickly, and we
+   expect it will continue to do so for some time. As a consequence, we
+   anticipate that packages will frequently be updated simply to take advantage
+   of new features. Optimizing for an ecosystem where everyone can reliable live
```
Comment (BasThomas) [inline proposals/NNNN-swiftpm-version-pinning.md]: reliable > reliably

File: proposals/NNNN-swiftpm-version-pinning.md [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+the resulting language consistent and beautiful.
+
+We have found significant consensus that without considering the prededent, the
+"lock" terminology is conceptually the *wrong* word for the operation being
+performed here. We view pinning as a workflow-focused feature, versus the
+specification in the manifest (which is the "requirement"). The meaning of pin
+connotes this transient relationship between the pin action and the underlying
+dependency.
+
+In constrast, not only does lock have the wrong connotation, but it also is a
```
Comment (BasThomas) [inline proposals/NNNN-swiftpm-version-pinning.md]: constrast > contrast

**Comment by ddunbar** [comment]
@BasThomas updated, thanks!

**Pull request merged**